server/src/main/java/org/elasticsearch/action/admin/cluster/state/TransportClusterStateAction.java

```diff
@@ -115,6 +116,13 @@ public void onTimeout(TimeValue timeout) {
         }
     }
 
+    // The cluster state can be a large object so we resolve the response listener on the MANAGEMENT pool to not serialize the resulting
+    // large response on a transport or cluster state thread which we don't want to block needlessly
+    private void resolveListener(ClusterStateRequest request, ClusterState currentState, ActionListener<ClusterStateResponse> listener) {
+        threadPool.executor(ThreadPool.Names.MANAGEMENT).execute(
+                ActionRunnable.supply(listener, () -> buildResponse(request, currentState)));
```

Comment:
Are you sure that `buildResponse()` is the expensive bit? That's not where the serialisation happens, it's just reorganising some maps a bit.

Reply:
It's not but the thread that resolves the listener is also the thread that will do the serialization in the TCP transport (`OutboundHandler`). I could obviously just resolve the listener on the management thread for that reason, but if I already fork-off then I might as well fork off as much work as I can?

Reply:
Right, sorry, I lost track.